오픈 마켓 [STEP 1] Yun, Nala, Marco

Marcorable opened the pull request:
@lina0322 

@keeplo @jazz-ing @blanche37

안녕하세요 엘림🤠

내용이 생각보다 어렵고 고민을 많이해서 늦어졌습니다!
불금인데... 저녁에 보내서 죄송해요...

저희 생각에 POST, PATCH, DELETE 
통신은 CollectionView를 보여주는 Step2 까지는 사용할 것 같지 않아서 더늦어질까봐 생략하고
UnitTest와 통신 컨트롤러(GET동작만) 구현해보았습니다.
이후에 필요할때 조금씩 수정해보려고 합니다..

## 고민되었던 점 해결하지 못한 부분 🤯
* **JSONDecoder 인스턴스 생성 위치**
아래코드와 같이 ParsingManager 라는 타입을 구현해서 JSONDecoding 작업이 필요할때마다
사용할 parse() 메서드를 구현해보았는데 decoder 인스턴스의 위치를 고민했습니다.
아래 두 경우의 큰 차이점을 잘 모르겠습니다. 어떤 키워드로 이런 차이를 고민하면 좋을지 궁금합니다..
~~~swift
// 구현된 코드
struct ParsingManager {
    let decoder = JSONDecoder()
    
    init() {
        decoder.keyDecodingStrategy = .convertFromSnakeCase
    }

    func parse<T: Decodable>(_ data: Data, to type: T.Type) -> Result<T, Error> {
        do {
            let result = try decoder.decode(type, from: data)
            return .success(result)
        } catch {
            return .failure(ParsingError.failDecoding)
        }
    }
}
// 고민한 코드
struct ParsingManager {
    func parse<T: Decodable>(_ data: Data, to type: T.Type) -> Result<T, Error> {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        do {
            let result = try decoder.decode(type, from: data)
            return .success(result)
        } catch {
            return .failure(ParsingError.failDecoding)
        }
    }
}
~~~

## 궁금한 점 🤔
* **의존성 주입?!**
사용자 타입을 구현하는 과정에서 아래와 같은 두가지 형태의 초기화에서 의존성 주입관련한 내용을
정확히 파악하지 못하고 있습니다. 아래와 같은 상황은 어떤 차이점이 있는지, 
의존성 주입관련한 어떤 특징이 있는지 궁금합니다
~~~swift
// 구현 코드
struct NetworkingManager {
    let session: URLSessionProtocol
    let parsingManager: ParsingManager
    let baseURL: String
    
    init(session: URLSessionProtocol, parsingManager: ParsingManager, baseURL: String) {
        self.session = session
        self.parsingManager = parsingManager
        self.baseURL = baseURL
    }
}
// 궁금한 코드
struct NetworkingManager {
    let session: URLSessionProtocol = URLSession.shared
    let parsingManager: ParsingManager = ParsingManager()
    let baseURL: String = "https://camp-open-market-2.herokuapp.com"
}
~~~

## 피드백 받고 싶은 부분 😹
* **RequestAPI 아이디어**
열거형으로 다양한 정보를 정리하기 힘들것 같고 추후 앱에 추가적인 통신 상황이 생길 경우 확장성을 고려해서 
하나의 타입을 생성하고 통신상황의 정보(path, httpBody, httpMethode)를 담았습니다.
이런 정보 묶음에 구현에 문제는 없는지 피드백 받고 싶습니다.
~~~swift
struct RequestAPI {}

//RequestAPI 인스턴스(let api)를 앱에서 가지고 있다가 (통신할 종류만큼) -> URLRequest에 정보 넣기

let request: URLRequest = manager.configureRequest(from: api)
~~~

* **추상화의 사용**
GET 메서드 통신에 한해서 ItemBundle을 받는 경우와 Item 을 받는 경우 두가지가 있는데
같은 메서드를 이용하기 위해서 탈출 클로저에 전달되는 Result 성공케이스 인자로 두가지 타입을 모두 수용하기 위해서
ResultArgument Protocol 으로 추상화를 해주었습니다. 
크게 기능은 없고 단순히 추상화 목적으로 구현해 보았는데 괜찮은 아이디어인지 피드백받고 싶습니다.
~~~swift 
protocol ResultArgument { }
class Item: Decodable, ResultArgument {}
class ItemBundle: Decodable, ResultArgument {}
func request(bundle request: URLRequest, completion: @escaping (Result<ResultArgument, Error>) -> ())
~~~

* **유닛테스트 비교방법**
Decodable을 채택한 두 모델 Item과 ItemBundle은 
채택된 프로토콜 때문에 원하는 init()을 구현해서 단순한 Dummy 인스턴스를 만드는게 어려웠습니다.
하드 코딩으로 기대되는 정보 값을 일일이 비교하는 형태로 디코딩된 인스턴스 프로퍼티 정보들과 비교해보았는데..
이런 방식이 나쁘지 않은지 아니면 더 좋은 방법은 없는지 궁금합니다!!
~~~swift
func test_success_JSON파일의정보와_디코딩된인스턴스정보가같다() {
        //given
        let path = Bundle(for: type(of: self)).path(forResource: "Item", ofType: "json")
        let jsonFile = try? String(contentsOfFile: path!).data(using: .utf8)
        let id = 1
            // ...
        //when
        let result = decoder.parse(jsonFile!, to: Item.self)
        guard case .success(let instance) = result else {
            XCTAssert(false)
            return
        }
        //then
        if instance.id == id {  // 다른 정보 생략
            XCTAssert(true)
        } else {
            XCTAssert(false)
        }
    }
~~~

------

엘림이 조언해주신 MockURLSession 구현이나 해당 방식이 올바르게 구현 되었는 지 
또, UnitTest에서 적절하게 테스트를 잘 사용한건지 궁금합니다!

저희가 ViewController에 실제 통신이 되는지도 테스트해본 테스트 코드를 주석처리해 두었습니다.(확인용)

blanche37 committed:
feat: Item 모델 타입 구현

상품정보를 보여 줄 모델
Decodable 채택
CodingKeys 선언

jazz-ing committed:
feat: ItemListPage 모델 타입 구현

page당20개씩 인스턴스를 생성하기 위한 타입
Decodable 채택
CodingKeys 선언

Marcorable committed:
refactor: 모델 클래스 이름변경 및 폴더 정리

ItemListPage 클래스 이름변경 -> ItemBundle
ItemBundle.itemList 프로퍼티 이름변경 -> ItemBundle.items
Model, SupportingFiles, View 폴더 생성 및 파일정리

blanche37 committed:
feat: 의존성 주입을 위한 URLSessionProtocol 프로토콜 생성

OpenMarketUnitTests생성

jazz-ing committed:
test: 테스트를 위한 MockURLSession과 MockURLSessionDataTask 구현

테스트를 위해 요청이 성공한 경우와 실패한 경우의 케이스를 나눠줌.

Marcorable committed:
feat: JSON 파일 추가 및 추가된 JSON 파일불러오기 구현

item.json, items.json 파일 UnitTests 폴더에 추가
MockURLSession().dataTask 메서드 내부에서 json 파일 읽고 Data 타입으로
반환하는 기능 구현

Marcorable committed:
feat: JSON 파일 파싱작업할 ParsingManager 타입 구현

convertFromSnakeCase 속성 추가한 초기화 추가
주어진 Data 와 원하는 T.Type으로 Decoding 하는 parse 메서드 구현

blanche37 committed:
test: ParsingManager타입 parse메서드 test

예상하는 결과 값과 파싱된 인스턴스 프로퍼티랑 일치하는지 테스트

jazz-ing committed:
refactor: JSONDecoder Case Style 속성 추가

jazz-ing committed:
feat: NetworkingManager 타입 및 request 메서드 구현

통신을 담당할 컨트롤러인 NetworkingManager
통신을 담당하는 메서드인 request()

blanche37 committed:
feat: RequestAPI 타입 구현

request시 전달 할 정보를 담아두는 타입

Marcorable committed:
feat: URLRequest 인스턴스를 생성해줄 configureRequest 메서드 구현

RequestAPI 인스턴스의 저장된 정보를 받아서 URLRequest 인스턴스를
생성해줄 매서드 구현

jazz-ing committed:
test: JSON 파싱 모든 케이스 테스트

blanche37 committed:
test: request 메서드 테스트 케이스 구현

lina0322 commented on OpenMarket/OpenMarket/Controller/NetworkingManager.swift:
```diff
@@ -0,0 +1,55 @@
+    
+    func configureRequest(from api: RequestAPI) -> URLRequest {
+        guard let url = URL(string: baseURL + api.path) else {
+            fatalError("URL 생성실패")
```
여기서 앱이 종료되는게 올바른 실행 방향일까요..🥲
 앱이 갑자기 꺼진다면 사용자는 앱에 대해서 어떻게 평가를 할까요?! 
좀 더 나은 에러처리는 없을까요?

jazz-ing replied:
에러 처리에 대해 일단은 생각하지 않고 코드를 작성해서 생긴 문제같습니다.
NetworkingManagerError라는 enum을 만들어서 각 케이스별로 에러 처리를 하도록 수정해주었습니다!

lina0322 commented on OpenMarket/OpenMarket/Controller/NetworkingManager.swift:
```diff
@@ -0,0 +1,55 @@
+        return request
+    }
+    
+    func request(bundle request: URLRequest, completion: @escaping (Result<ResultArgument, Error>) -> ()) {
```
Result 사용 좋습니다. 💯
Result는 어떻게 구성되어 있을까요?🧐
(ㅈㅏ, 여기 코드 담당자 아니신분 키보드에 손 좀 올려볼까여.)

jazz-ing replied:
- Result는 success와 failure라는 case를 가진 enum이라고 알고 있습니다!
각 케이스의 연관값을 제네릭으로 받는데 success의 경우 원하는 결과의 타입을, failure의 경우 Error타입을 받을 수 있도록 구현해놓았다고 알고있습니다 :)
- `request` 메서드에서 Result를 사용한 이유는 통신의 결과를 확인해 통신이 성공한 경우, 실패한 경우를 나눠서 상황에 맞게 처리하기 위해서였습니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/NetworkingManager.swift:
```diff
@@ -0,0 +1,55 @@
+    func request(bundle request: URLRequest, completion: @escaping (Result<ResultArgument, Error>) -> ()) {
+        let dataTask = session.dataTask(with: request) { data, response, error in
+            guard error == nil else {
+                completion(.failure(error!))
+                return
+            }
+            guard let data = data,
+                  let response = response as? HTTPURLResponse,
+                  (200..<300).contains(response.statusCode) else {
+                completion(.failure(error!))
```
위에 에러와 이곳의 에러의 강제 언래핑을 보는 제 마음이 조금 아픈데, 방법이 없을까유? 😭
추가로 지금 `(200..<300).contains(response.statusCode)` 부분은 어떤걸 처리하고 있는지 설명해줄 수 있을까요?!
(해당 코드를 작성한분 말고! 다른분이 설명해주시면 좋겠습니다~!)

jazz-ing replied:
- 이 부분에 대해서도 NetworkingManagerError enum으로 에러처리 구현을 해주었습니다~!
- `(200..<300).contains(response.statusCode)` 해당 코드에 대해서는 서버에서 주는 response에 대해 성공하는 상태코드와 아닌 경우를 구분해주기 위해 작성해보았습니다!

lina0322 commented on OpenMarket/OpenMarket/Controller/NetworkingManager.swift:
```diff
@@ -0,0 +1,55 @@
+                completion(.failure(error!))
+                return
+            }
+            let parsedData = parsingManager.parse(data, to: ItemBundle.self)
```
여기서 ItemBundle로 타입이 지정되어있다면, 만들어 놓은 ResultArgument를 추상화한 의미가 있을까요...?!😯

Marcorable replied:
후행 클로저로 디코딩 완료된 인스턴스를 전달할 때 필요하다고 생각해서 구현 했는 데, 
디코딩 작업이 `request()` 메서드의 후행 클로저에서 동작하는 게 좀 더 좋은 방향일 것 같아서 
추상화를 없애는 방향으로 수정했습니다!

lina0322 commented on OpenMarket/OpenMarket/Controller/ParsingManager.swift:
```diff
@@ -0,0 +1,29 @@
+
+import Foundation
+
+enum ParsingError: Error {
+    case failDecoding
+}
```
ParsingManager에서만 사용할 거라면, 내부에서 가지고 있어도 되지 않을까요?🤓

blanche37 replied:
ParsingManager에서 만 사용되고 있어, 내부에 선언하였습니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/ParsingManager.swift:
```diff
@@ -0,0 +1,29 @@
+}
+
+struct ParsingManager {
+    let decoder = JSONDecoder()
```
외부에서 접근해야 하는 값일까요?😏

blanche37 replied:
아닌 것 같아, 접근제어자 추가하였습니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/ParsingManager.swift:
```diff
@@ -0,0 +1,29 @@
+    let decoder = JSONDecoder()
+    
+    init() {
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
```
오! convertFromSnakeCase를 사용하다니👏

lina0322 commented on OpenMarket/OpenMarket/Model/Dependency/ResultArgument.swift:
```diff
@@ -0,0 +1,10 @@
+
+import Foundation
+
+protocol ResultArgument { }
```
근데 이거 채택하는 친구들이, Decodable이라는 공통점을 가지고 있지 않나요?😶

Marcorable replied:
매우 그렇게 생각합니다.. 근데 고민해보니 추상화가 필요 없을 것 같아서 
저 친구들은 그대로 두려고 합니다.......😭

lina0322 commented on OpenMarket/OpenMarket/Model/Dependency/URLSessionProtocol.swift:
```diff
@@ -0,0 +1,14 @@
+//  Created by yun on 2021/08/17.
+//
+
+import Foundation
+
+protocol URLSessionProtocol {
+    func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
+}
+
+extension URLSession: URLSessionProtocol { }
```
LGTM💚

jazz-ing replied:
GOTCHA🧡

lina0322 commented on OpenMarket/OpenMarket/Model/Item.swift:
```diff
@@ -0,0 +1,40 @@
+
+import Foundation
+
+class Item: Decodable, ResultArgument {
```
해당타입과  ItemBundle을 struct가 아니라 class를 선택하게된 이유가 궁금해요!

Marcorable replied:
처음 생각은 List 보여주는 화면에 조회된 item(8개 정보 가진)을 조회 화면에 참조시켜서 (10개 정보) 업데이트 시켜주는 목적을 가졌었는데. 고민해보니 리스트의 새로고침을 위해 매번 통신을 해서 업데이트 해주고 굳이 참조시킬필요 없다고 생각해서 구조체로 선택했습니다.

Marcorable replied:
Date 인스턴스를 프로퍼티로 갖기 위해 현했던 초기화 부분도 앱에서 사용하는 일이 따로 없는것 같아서 제거했습니다!

lina0322 commented on OpenMarket/OpenMarket/Model/Item.swift:
```diff
@@ -0,0 +1,40 @@
+    var images: [String]?
+    let registrationDate: Date
+    
+    private enum CodingKeys: String, CodingKey {
+        case id, title, descriptions, price, currency, stock, thumbnails, images, discountedPrice, registrationDate
+    }
```
`decoder.keyDecodingStrategy = .convertFromSnakeCase`를 사용했지만 CodingKeys를 작성한 이유는 required init 때문 인가욥..?!

jazz-ing replied:
넵 그렇습니다!

lina0322 commented on OpenMarket/OpenMarket/Model/Item.swift:
```diff
@@ -0,0 +1,40 @@
+        case id, title, descriptions, price, currency, stock, thumbnails, images, discountedPrice, registrationDate
+    }
+    
+    required init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.id = try container.decode(Int.self, forKey: .id)
+        self.title = try container.decode(String.self, forKey: .title)
+        self.descriptions = try? container.decode(String.self, forKey: .descriptions)
+        self.price = try container.decode(Int.self, forKey: .price)
+        self.discountedPrice = try? container.decode(Int.self, forKey: .discountedPrice)
+        self.currency = try container.decode(String.self, forKey: .currency)
+        self.stock = try container.decode(Int.self, forKey: .stock)
+        self.thumbnails = try container.decode([String].self, forKey: .thumbnails)
+        self.images = try? container.decode([String].self, forKey: .images)
+        let epochTime = try container.decode(Double.self, forKey: .registrationDate)
+        self.registrationDate = Date(timeIntervalSince1970: epochTime)
+    }
```
required init에 decoder container를 생성하였네요! container는 어떤 이유 때문에 만드셨는지와, 요 코드가 생기게된 과정도 궁금해요👀

jazz-ing replied:
Date 타입으로 초기화하는 부분 때문에 구분을 하려고 구현했습니다. `init()` 내부에서 `registrationDate` 를 Double에서 Date 타입으로 변환해 프로퍼티로 저장하고 있습니다~!

lina0322 replied:
오, 그런데 혹시 연산프로퍼티가 어떤건지 알고계시나요?!

Marcorable replied:
저장 프로퍼티에 JSON 데이터를 바로 저장하고(숫자) 연산 프로퍼티를 이용해서 Date 타입으로 변환해주는 방법도 고민을 해보았는데요,
만약 두 방법 중 고른다면 어떤 방법이 더 좋을까요? 
저희가 생각한 이유중에 하나는 JSON 파일에서 기대되지 않는 타입의 프로퍼티를 가진 Swift 내부 사용자 타입을 협업하는 동료나 다른 파트를 작업하는 사람이 파악하기 어렵기 때문에 본래 타입을 유지하는건 아닌가 하는 이유도 생각해봤구요. 
또 하나의 이유는 다양한 방향으로 해석?!처리?! 될 수 있는 데이터를 디코딩(초기화) 과정에서 Date 타입 형태로 정해버려서 확장성을 막아버릴 수 있는 이유가 아닐까 하는 생각도 해보았습니다. 
실제로 어떻게 많이 쓰는지 알고 싶습니다.

lina0322 commented on OpenMarket/OpenMarket/Model/ItemBundle.swift:
```diff
@@ -0,0 +1,18 @@
+    let page: Int
+    var items: [Item]
+    
+    enum CodingKeys: String, CodingKey {
```
Item처럼 요기 코딩키에도 private해주면 오때여

jazz-ing replied:
> Item처럼 요기 코딩키에도 private해주면 오때여

jazz-ing replied:
이 부분은 프로퍼티 이름과 json 키가 동일해서 삭제했습니다 :)

lina0322 replied:
저보다 낫네요..😆

lina0322 commented on OpenMarket/OpenMarket/Model/RequestAPI.swift:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+    let method: HTTPMethod
+    let path: String
```
path가 string으로 외부에서 하드코딩으로 주입 되었을때 위험한 것은 없을까요?🤔

blanche37 replied:
String을 다루는 OpenMarketInfo 열거형을 추가해봤습니다.

lina0322 commented on OpenMarket/OpenMarket/Model/RequestAPI.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+    let method: HTTPMethod
+    let path: String
+    let body: Data?
```
오 여기서 body를 갖고있네여. 앞으로 코드진행이 어떻게 될지 궁금합니다ㅎㅎ

lina0322 commented on OpenMarket/OpenMarketUnitTests/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,57 @@
+            fatalError()
+        }
+        
+        mockURLSessionDataTask.resumeDidCall = { completionHandler(data, self.isSuccess ? success : failure, nil)}
```
실패해도 데이터는 전달되는 URLSession 이네욤ㅎㅎ (그냥 그렇다구용..😂)
그리고 저기 맨 뒤에 띄어쓰기 하나만 해줬으면 좋겠다고 생각해써옇ㅎㅎㅎㅎ

blanche37 replied:
성공시에는 데이터를 전달하고, 실패시에는 nil을 전달하도록 수정하였습니다.

lina0322 commented on OpenMarket/OpenMarketUnitTests/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,57 @@
+        }
+        
+        guard let fileName = DummyURL(rawValue: request.url!.absoluteString)?.fileName else {
+            fatalError("URL 이상함")
+        }
+        
+        let path = Bundle(for: type(of: self)).path(forResource: fileName, ofType: "json")
+        let jsonData = try? String(contentsOfFile: path!).data(using: .utf8)
+        
+        guard let data = jsonData else {
+            fatalError()
```
저는 fatalError가 요기 있는게 이상하다고 생각해요😸
그리고 혹시, 지금 이 코드에서 잘못된 URL을 가지고있는 리퀘스트가 들어오면 터지는건가요?

blanche37 replied:
fatalError()대신 실패하는 completionHandler를 호출하는 것으로 변경하였습니다.

lina0322 commented on OpenMarket/OpenMarketUnitTests/Mock/MockURLSessionDataTask.swift:
```diff
@@ -0,0 +1,16 @@
+//
+//  Created by 이윤주 on 2021/08/17.
+//
+
+import Foundation
+@testable import OpenMarket
+
+class MockURLSessionDataTask: URLSessionDataTask {
+    var resumeDidCall: () -> () = { }
+    override func resume() {
+        resumeDidCall()
+    }
+}
```
여기에 왜 resumeDidCall()이 필요할까요👉👈
왜 resume을 override하고있을까요?
(해당 코드를 작성하지 않은 분이 설명해주셨으면 좋겠어용ㅎㅎ)

blanche37 replied:
URLSession()의 dataTask()메서드의 completionHandler는 resume()메서드가 호출된 이후 실행되는데, 테스트를 위한 MockURLSessionDataTask의 메서드에서도 resume()이 호출된이후에 실패,성공 케이스에 따른 completionHandler를 실행하기위해서, resumeDidCall에 completionHandler를 참조해줬습니다.

blanche37 replied:
통신 없이 completionHandler를 호출하기 위해서 resume을 override 했습니다.

lina0322 commented on OpenMarket/OpenMarketUnitTests/OpenMarketUnitTests.swift:
```diff
@@ -0,0 +1,95 @@
+    var mockSesssion: URLSessionProtocol = MockURLSession()
+    var manager: NetworkingManager = NetworkingManager(session: MockURLSession(), parsingManager: ParsingManager(), baseURL: "https://camp-open-market-2.herokuapp.com")
+    
+    func test_success_JSON파일의정보와_디코딩된인스턴스정보가같다() {
```
성공하는 경우라고 되어있는데... 저는 실패가 뜨네요... 왜지..?!😿 
추가로 MockURLSession에 isSuccess이 존재하는데, 요게 false일때 테스트도 있으면 좋겠네요! 
마찬가지로, item일때와 items일때 테스트도 가능하겠네요?! 와우!🎉
(요기 유닛 테스트 담당 안 하신분이 새롭게 도전하면 좋은 경험이 될지도..?)

jazz-ing replied:
- 저희가 만든 custom json decoder인 `parse()`메서드는 제네릭으로 구성되어있는데요. 제네릭이 잘 작동하는지 확인해보기 위해  Item.json과 Items.json 파일을 이용해 두 타입 모두 디코딩이 잘 되는지 확인하는 테스트도 추가했습니다.
- isSuccess를 이용해서 통신이 성공했을 때와 실패했을 때를 나누어 테스트하도록 추가해보았습니다!

lina0322 reviewed (changes requested):
안녕하십니까, 세 분!🙇🏻‍♀️

여러분이 회의실에서 2주간 고구분투하는 모습을 보면서 제 캠프 생활이 떠올랐습니다...🥲
열심히하는 모습 정말 보기 좋아요!

여러분의 코드의 완성도 보다, 작성하는 동안 했던 고민들과 검색들, 수많은 대화들이 여러분을 더 멋지게 만들어 줄겁니다.🤘
늘 과정을 더 중요시 생각하는 여러분이 되시길 응원하며(????) 질문 먼저 답변하며 리뷰 시작하게씀다!

---
> 1. JSONDecoder 인스턴스 생성 위치
```swift
struct ParsingManager {
    func parse<T: Decodable>(_ data: Data, to type: T.Type) -> Result<T, Error> {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase // 이것과 윗줄의 위치를 고민중이시군요!
        do {
            let result = try decoder.decode(type, from: data)
            return .success(result)
        } catch {
            return .failure(ParsingError.failDecoding)
        }
    }
}
```
여러분 생각은 어떤가요? 지금처럼 parse에서만 사용한다면 어디에 있는게 좋을까요?! 또는 메소드가 더 늘어나서 다른 메소드에서도 필요하다면 어떨까요??

> 2. 의존성 주입
```swift
struct NetworkingManager { 
    // 이 세가지 프로퍼티의 의존성 주입이 고민되시나여?!
    let session: URLSessionProtocol = URLSession.shared
    let parsingManager: ParsingManager = ParsingManager()
    let baseURL: String = "https://camp-open-market-2.herokuapp.com"
}
```
우선 session은 어떻게 생각하시나요? 의존성이 없다면, 만들어 두신 테스트용 코드가 주입이 안될거같지않나요?! 요거는 뭔가 확실한 이유를 가지고 주입하게 하신 것 같은데,
나머지 두 개를 주입하게 만든 이유를 먼저 물어봐도 될까요?!

> 3. RequestAPI 아이디어

여러분의 아이디어니까! 뒤 스텝 진행하면서 어떤 장단점을 가지고 있는지 지켜봅시다:)

> 4. 추상화의 사용

좋은 시도라고 생각합니다! 관련해서 해당 부분에 피드백 남겼는데, 천천히 확인해보시죠.

> 5. 유닛테스트 비교방법
Decodable을 채택한 두 모델 Item과 ItemBundle은
채택된 프로토콜 때문에 원하는 init()을 구현해서 단순한 Dummy 인스턴스를 만드는게 어려웠습니다.
하드 코딩으로 기대되는 정보 값을 일일이 비교하는 형태로 디코딩된 인스턴스 프로퍼티 정보들과 비교해보았는데..
이런 방식이 나쁘지 않은지 아니면 더 좋은 방법은 없는지 궁금합니다!!

제가 이 질문을 제대로 이해했는지 모르겠는데요!
json파일에 있는 데이터를 확인하기 위해, 하드코어한 dummy를 만들어 비교했다는 말씀이실까요..?
지금처럼 모든 데이터가 맞는지 확인하기 위해서 이렇게 작업하는 것도 좋고,
id만 확인해보는 등 부분만 확인하는 것도 방법이 될 수 있겠죠?
그리고 이 방식도 나쁘지 않다고 생각합니다!

(확인을 위해 뷰컨에 주석으로 남겨주신것도 잘 돌려봤습니다.ㅎㅎ)

---

아래에 리뷰를 남겨두었으니, 천천히 확인해주십쇼!
세분이서 함께 고민해보시고, 의견 있는 것은 댓글 달아주세요.
(하나씩 달아도 되고, 한번에 모아서 달아주셔도 됩니당.)
따로 할말 없는 것은 이모지로 대신하셔도 충분합니다.

항상 말씀 드리지만, **정답은 없어요**. 
### 하지만 좋은 코드는 분명 있다고 생각합니다:)
우리 함께 좋은 코드가 무엇인지 고민해 보아요.💃

더불어 제가 놓친 부분이 있을 수도 있으니, 개인적으로는 시간날때 다른 캠퍼들이 받은 리뷰도 보면서 생각해 보길 추천드립니다.💪


갑자기 하고 싶은 말 있어서 추가해여!
세 분이 커밋을 번갈아가면서 한거 완전 보기좋아여!!💚

jazz-ing committed:
refactor: NetworkinManager 타입 에러 처리 기능 구현

blanche37 committed:
refactor: 접근제어자 추가, ParsingError위치변경

변수 저장프로퍼티에 접근제어자 추가
변수의 값에 접근하는 연산프로퍼티 추가

Marcorable committed:
refactor: 모델 추상화 제거 및 관련 코드 수정

두 모델을 하나로 묶어주던 ResultArgument 프로토콜 제거
ResultArgument로 업캐스팅되어 반환되던 처리부분 수정

jazz-ing committed:
refactor: 모델 프로퍼티 선언 방식 변경

Item 타입 - 값이 수정되지 않는 상수로 모두 변경
ItemBundle 타입 - 프로퍼티명과 JSON 키값이 동일해 CodingKeys 삭제

blanche37 committed:
refactor: MockURLSession 에러처리구현

Marcorable committed:
refactor: 두 모델의 타입 유형 변경

참조 타입인 class 타입에서 struct 타입으로 유형 변경함
Date 인스턴스로 변형해서 프로퍼티를 저장하기 위해 구현했던 초기화부분 삭제
Decodable 기본 구현으로 초기화가 처리되어있기 때문에 생략가능

jazz-ing committed:
test: 테스트 수정

각 테스트마다 테스트에 필요한 인스턴스 생성해 사용하는 방식으로 변경
parse 메서드가 제네릭하게 기능하는지 확인하기위해 Items.json 파일의 테스트 추가
MockURLSession의 isSuccess 프로퍼티를 통해 통신의 성공, 실패 여부를 결정하는 방식으로 테스트 수정

blanche37 committed:
feat: OpenMarketInfo 열거형 추가

하드코딩 될 수 있는 String들을 처리

Marcorable committed:
refator: UnitTest 에러 처리

lina0322 commented:
세 분 모두 수고 많으셨습니다! 

이번 리뷰에서 나왔던 이야기들도, 디스코드에서 나누었던 이야기들도 다 좋았습니다~
step1을 진행하면서 많은 것을 배우고 느끼셨을 것이라고 생각됩니다.
혹시라도 이해가 조금 안 됐던 부분이 있다면, 그냥 지나치지 않으시길 바랍니당!!

말씀하셨던 것 처럼(?) step2 빠르게 PR 보내주십셔!! 기대하고 있겠습니다.
추가적인 피드백은 step2 보면서 다시 해봅시다.

화이팅!🤜   🤛
